transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamRequestManager.java

```diff
@@ -0,0 +1,134 @@
+        new AddStreamRequest()
+            .streamType(clientStream.getStreamType())
+            .streamId(clientStream.getStreamId())
+            .metadata(clientStream.getMetadata());
+
+    servers.forEach(brokerId -> executor.run(() -> doAdd(request, brokerId, clientStream)));
+  }
+
+  private void doAdd(
+      final AddStreamRequest request,
+      final MemberId brokerId,
+      final ClientStream<M, P> clientStream) {
+    if (clientStream.isConnected(brokerId)) {
+      return;
+    }
+
+    final CompletableFuture<byte[]> result =
+        communicationService.send(
+            StreamTopics.ADD.topic(),
+            request,
+            BufferUtil::bufferAsArray,
+            Function.identity(),
+            brokerId,
+            REQUEST_TIMEOUT);
+
+    result.whenComplete(
+        (ignored, error) -> {
+          if (error != null) {
+            LOG.warn(
+                "Failed to open stream {} to node {}. Will retry in {}",
+                clientStream,
+                brokerId,
+                RETRY_DELAY);
+            // TODO: define some abort conditions. We may not have to retry indefinitely.
+            // For now we retry always. Eventually the request will succeed. Duplicate add request
+            // are fine.
+            executor.schedule(RETRY_DELAY, () -> doAdd(request, brokerId, clientStream));
+          } else {
+            LOG.debug("Opened stream {} to node {}", clientStream, brokerId);
+            clientStream.add(brokerId);
+          }
+        });
+  }
+
+  void removeStream(final ClientStream<M, P> clientStream, final Collection<MemberId> servers) {
+    final var request = new RemoveStreamRequest().streamId(clientStream.getStreamId());
+    servers.forEach(brokerId -> executor.run(() -> doRemove(request, brokerId, clientStream)));
+  }
+
+  private void doRemove(
+      final RemoveStreamRequest request,
+      final MemberId brokerId,
+      final ClientStream<M, P> clientStream) {
+
+    final CompletableFuture<byte[]> result =
+        communicationService.send(
+            StreamTopics.REMOVE.topic(),
+            request,
+            BufferUtil::bufferAsArray,
+            Function.identity(),
+            brokerId,
+            REQUEST_TIMEOUT);
+
+    result.whenCompleteAsync(
+        (ignored, error) -> {
+          if (error != null && clientStream.isConnected(brokerId)) {
+            // TODO: use backoff delay
+            executor.schedule(RETRY_DELAY, () -> doRemove(request, brokerId, clientStream));
+          } else {
+            LOG.debug("Removed stream {} to node {}", clientStream, brokerId);
+            clientStream.remove(brokerId);
+          }
+        },
+        executor::run);
+  }
+
+  void removeAll(final Collection<MemberId> servers) {
+    servers.forEach(this::doRemoveAll);
+  }
+
+  private void doRemoveAll(final MemberId brokerId) {
+    // Do not wait for response, as this is expected to be called while shutting down.
+    communicationService.send(
```

Comment:
:wrench: In this case, we can use `unicast`. If we need TCP, we can pass the last argument as `true`. Though if it's best of effort, maybe UDP is fine? :shrug: